src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -716,17 +720,18 @@ private void CreateGrainInstance(string grainTypeName, ActivationData data, stri
             {
                 Grain grain;
 
+                data.ServiceScope = this.serviceScopeFactory.CreateScope();
                 //Create a new instance of the given grain type
-                grain = grainCreator.CreateGrainInstance(grainType, data.Identity);
+                grain = grainCreator.CreateGrainInstance(data.ServiceProvider, grainType, data.Identity);
 
                 //for stateful grains, install storage bridge
                 if (grain is IStatefulGrain)
                 {
                     SetupStorageProvider(grainType, data);
 
                     var storage = new GrainStateStorageBridge(grainType.FullName, data.StorageProvider);
-
-                    grain = grainCreator.CreateGrainInstance(grainType, data.Identity, stateObjectType, storage);
+                    // TODO: avoid double creation!
+                    grain = grainCreator.CreateGrainInstance(data.ServiceProvider, grainType, data.Identity, stateObjectType, storage);
```

Comment:
yes, I can certainly do that. I still wanted to have a perf run with the same allocation to compare apples to apples. But on further thought, I don't think that's worse it, nor it preserves the apples (as it uses the scoped SP to build the instance that will be discarded, and I don't know if it's as cheap as before, probably it is, but it's no reason to not fix it now).